### ISSUE TITLE
Ctrl-Q/Ctrl-S no longer works during packer

### ISSUE THREAD

**Issue by koitsu** [hashicorp/packer#8222]
At some point between Packer 1.3 and 1.4, something changed where Ctrl-Q (a.k.a. XON) and Ctrl-S (a.k.a. XOFF) no longer work from an interactive terminal whilst packer runs.  The below was confirmed in both Packer 1.4.4 and Packer 1.4.1.  I'm pretty sure 1.3.x did not have this problem but can't remember; I am CERTAIN 1.2.x did not.

For those unfamiliar with what I'm talking about: these are terminal/tty control features for flow control, which permit pause and resumption of terminal output, respectively. You might know them better as Ctrl-Q (`start`) and Ctrl-S (`stop`) or from serial port flow control days.

It appears that packer, or some Go module (or maybe even newer Go itself) is doing this intentionally.  Proof is below, where the terminal in question using bash + packer is `/dev/pts/0`. Both FreeBSD 11.3 (stable/11) 64-bit and Ubuntu Linux 18.04.3 LTS (bionic) 64-bit were tested.  We can check the tty params of another tty using `stty -f` on BSD or `stty -F` on Linux, alongside the `-a` argument:

FreeBSD: prior to packer running:
```
$ stty -a -f /dev/pts/0 | grep ixon
...
iflags: -istrip -icrnl -inlcr -igncr ixon -ixoff ixany imaxbel -ignbrk
```
FreeBSD: whilst packer is running:
```
$ stty -a -f /dev/pts/0 | grep ixon
...
iflags: -istrip -icrnl -inlcr -igncr -ixon -ixoff ixany imaxbel -ignbrk
```

Linux: prior to packer running:
```
$ stty -a -F /dev/pts/0 | grep ixon
-ignbrk -brkint -ignpar -parmrk -inpck -istrip -inlcr -igncr -icrnl ixon -ixoff
```
Linux: whilst packer is running:
```
$ stty -a -F /dev/pts/0 | grep ixon
-ignbrk -brkint -ignpar -parmrk -inpck -istrip -inlcr -igncr -icrnl -ixon -ixoff
```

Many find terminals "confusing", so please understand that XON/XOFF do not directly correlate with the flags you see called `ixon` or `ixoff`. These control input vs. output handling of said XON/XOFF characters and are separate flags (bits) within terminal structurse. `ixon` pertains to the ability to manage output flow control (e.g. pause/resume output of data to a terminal screen).

On both OSes, note that `ixon` becomes `-ixon`, hence Ctrl-Q and Ctrl-S keys are ignored. Something code-wise is changing the tty params via ioctls. Most likely this is being done through termios / `tcXXX` calls.

On FreeBSD, this would be caused by setting bit 0x200 of `c_iflag` of a `termios` struct to 1.  See `#define IXON` per `sys/_termios.h` alongside `termios(4)` man page.

On Linux, this would be caused by setting bit 0x2000  of `c_iflag` of a `termios` struct to 1.  See `#define IXON` per `x86_64-linux-gnu/bits/termios.h` or `asm-generic/termbits.h` (unsure which) alongside `termios(3)` man page.

Sanity is restored once packer exits.

If this is a side-effect of some other module or thing: can we find out what is doing it, why it is doing it, and how to restore sane behaviour that admins have depended on since the early 70s (likely introduced in AT&T UNIX version 1, possibly 2).

If this was done intentionally: why?  It is extremely common for systems administrators, during heavy terminal output, to press Ctrl-S to pause (buffer up) the output so that they can scroll back safely and review what they just saw. With today's terminal emulators, this is often done to keep the terminal emulator from scrolling to the bottom when receiving data. This is even documented in the [FreeBSD termios(4) man page](https://www.freebsd.org/cgi/man.cgi?query=termios&sektion=4) (search for "fast terminals"). The buffering I described happens deeper within the tty/pty code and possibly parts of libc (for stdout buffering) and it is not a concern of the userland program (ex. packer) -- kernel, pty/tty, and libc subsystems take care of the rest.  It is not always feasible to insist that a user use `| tee -a output.log` and then sift through that post-run just to catch output they just saw.

Thank you!

**Comment by SwampDragons** [comment]
Thanks for reaching out. I can reproduce. This probably happened as part of the modifications we made when adding the progress bar. We've had a number of unintended consequences to that UI enhancement 😬 

I'll try to take a look when I can.

**Comment by SwampDragons** [comment]
Pretty sure this is caused by https://github.com/hashicorp/packer/pull/7352; control flow works as expected in v1.3.5 and is broken in v1.4.0.

**Comment by SwampDragons** [comment]
Just as an update; I've made a PR to the tty library we use to address this. If that doesn't go through, we'll investigate how to fix this from Packer's side.

**Comment by koitsu** [comment]
@SwampDragons If it would be at all helpful, I can try to build packer from modified source (assuming it's possible via something like `go get` but with `git revert 6e442e6` applied somewhere within the tree in advance) to verify that's the PR which introduced it.  While I'm an SA by profession, I don't work much with Go, only have general familiarity with how to wrangle Go-based things.  I'd also need to know what version of Go HashiCorp uses to build Packer FreeBSD binaries (1.13.3 is what stable/11 FreeBSD offers natively right now.  Happy to try/do the same on Linux, just that it's easier for me on FreeBSD.)

**Comment by SwampDragons** [comment]
I appreciate it, but I've already been able to reproduce on OSX and verify that my original guess was right and #7352 was the source.  I think this is an issue for all unix flavors.

I've already tested https://github.com/mattn/go-tty/pull/27  and verified that it will solve our issues if the maintainers of that library merge it. If they don't, I have a couple of ideas for further changes we can make to try to retain appropriate control flow behaviors, but this is the path of least resistance.
